### PR TITLE
feat(sdk): detect stale nodes

### PR DESCRIPTION
**Pull request by lklimek**
## Issue being fixed or feature implemented

When fetching information from a node that has chain halt, we receive outdated information.

## What was done?

Implemented two mechanisms to detect stale nodes:

1. Ensure height of received response  is not older than the last seen height minus some tolerance (defaults to 1).
2. Ensure time of received response is within some time window (mechanism disabled by default).
3. Added error StaleNodeError to Sdk  that is emitted by Sdk if the conditions above are not met. 
4. Added CanRetry trait to dash_sdk::Error that will return true is the request should be retried

## How Has This Been Tested?

Added tests.

## Breaking Changes

None on API level, but the behavior changed a bit.

`rs_dapi_client::CanRetry::is_node_failure()` is deprecated in favor of `can_retry()`.

## Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have added "!" to the title and described breaking changes in the corresponding section if my code contains any
- [ ] I have made corresponding changes to the documentation if needed

**For repository code-owners and collaborators only**
- [ ] I have assigned this pull request to a milestone

## Summary by CodeRabbit

- **New Features**
	- Enhanced error handling with a new `StaleNodeError` type for detailed reporting of stale metadata issues.
	- Added retry logic improvements in the DAPI client, allowing for more nuanced error handling.
	- Introduced new fields in the SDK for managing proof staleness based on height and time.
	- Added methods for verifying proof metadata against defined tolerances.

- **Bug Fixes**
	- Updated logic for determining retry capabilities based on error types, improving reliability during DAPI requests.

- **Documentation**
	- Improved clarity in documentation comments related to the new retry logic and error handling.

- **Tests**
	- Added unit tests for new verification functions to ensure proof validity checks work as intended.

### CODE REVIEW

File: packages/rs-dapi-client/src/dapi_client.rs [hunk 1]
```diff
@@ -264,12 +264,12 @@ impl DapiRequestExecutor for DapiClient {
                     duration.as_secs_f32()
                 )
             })
-            .when(|e| e.is_node_failure())
+            .when(|e| !e.can_retry().unwrap_or(false))
```
[the remark this answers is not in the document]
Comment (lklimek) [reply packages/rs-dapi-client/src/dapi_client.rs:267]: Not retrying on unspecified (None) state is safe choice. In future, we might improve implementations of CanRetry trait to inspect errors in more details, and return None in less cases than now.

File: packages/rs-drive-proof-verifier/src/error.rs [hunk 2]
```diff
@@ -82,6 +82,37 @@ pub enum Error {
...
     ContextProviderError(#[from] ContextProviderError),
+
+    /// Proof is stale; try another server
+    #[error(transparent)]
+    StaleProof(#[from] StaleProofError),
+}
+
+/// Received proof is stale; try another server
+#[derive(Debug, thiserror::Error)]
+pub enum StaleProofError {
```
Comment (shumkov) [inline packages/rs-drive-proof-verifier/src/error.rs:93]: Are you sure we want to do that on the verifier side? Shouldn't be this logic located in SDK so we can guarantee freshness for all types of queries that have metadata?

File: packages/rs-sdk/src/sdk.rs [hunk 3]
```diff
@@ -260,7 +287,7 @@ impl Sdk {
             .context_provider()
             .ok_or(drive_proof_verifier::Error::ContextProviderNotSet)?;
 
-        match self.inner {
+        let (object, mtd, proof) = match self.inner {
```
Comment (shumkov) [inline packages/rs-sdk/src/sdk.rs:290]: Let's don't use shortcuts. It will be confusing for other devs

File: packages/rs-sdk/src/sdk.rs [hunk 4]
```diff
@@ -544,6 +574,108 @@ impl Sdk {
...
+}
+
+/// If current proof height is behind previous proof height by more than `tolerance`, the proof is considered stale.
+/// This is used to verify the freshness of the proof.
+fn verify_proof_height(
+    metadata: &ResponseMetadata,
+    tolerance: u64,
+    previous_proof_height: Arc<atomic::AtomicU64>,
+) -> Result<(), drive_proof_verifier::Error> {
+    let mut prev = previous_proof_height.load(Ordering::Relaxed);
```
Comment (shumkov) [inline packages/rs-sdk/src/sdk.rs:624]: shortcut

File: packages/rs-sdk/src/sdk.rs [hunk 5]
```diff
@@ -544,6 +574,108 @@ impl Sdk {
...
+        return Ok(());
+    }
+
+    // If received proof height is behind previous proof height by more than PROOF_HEIGHT_TOLERANCE, the proof is stale.
+    //
+    // If prev is less than tolerance, then Sdk just started, so we just trust the proof, assuming we connected to a
+    // trusted node.
+    // FIXME: in future, we need to securely trust some node (like a seed) from which we will fetch list of nodes and
+    // initial height.
+    if prev > tolerance && received < prev - tolerance {
```
Comment (shumkov) [inline packages/rs-sdk/src/sdk.rs:644]: add height postfix to these variables

File: packages/rs-dapi-client/src/lib.rs [hunk 6]
```diff
@@ -71,9 +71,19 @@ impl<T: transport::TransportRequest + Send> DapiRequest for T {
     }
 }
 
-/// Allows to flag the transport error variant how tolerant we are of it and whether we can
-/// try to do a request again.
+/// Returns true if the operation can be retried.
```
Comment (shumkov) [inline packages/rs-dapi-client/src/lib.rs:74]: I think having the third state None doesn't provide additional benefits but makes it more complex.

**Review by shumkov — CHANGES_REQUESTED** [review]

File: packages/rs-dapi-client/src/lib.rs (hunk 6 above)
Comment (lklimek) [reply packages/rs-dapi-client/src/lib.rs:74]: done

File: packages/rs-sdk/src/sdk.rs (hunk 3 above)
Comment (lklimek) [reply packages/rs-sdk/src/sdk.rs]: done

File: packages/rs-drive-proof-verifier/src/error.rs (hunk 2 above)
Comment (lklimek) [reply packages/rs-drive-proof-verifier/src/error.rs]: renamed to StaleNodeError and moved to Sdk.

**Comment by lklimek** [comment]
@coderabbitai full review

File: packages/rs-sdk/src/error.rs [hunk 7]
```diff
@@ -80,3 +84,42 @@ impl From<PlatformVersionError> for Error {
...
     }
 }
+
+impl CanRetry for Error {
+    fn can_retry(&self) -> bool {
+        matches!(
+            self,
+            Error::StaleNode(..)
+                | Error::DapiClientError(_)
+                | Error::CoreClientError(_)
```
[the remark this answers is not in the document]
Comment (lklimek) [reply packages/rs-sdk/src/error.rs:94]: ok

File: packages/rs-sdk/src/sdk.rs [hunk 8]
```diff
@@ -544,6 +567,105 @@
...
+/// - `tolerance_ms`: Tolerance in milliseconds
+fn verify_metadata_time(
+    metadata: &ResponseMetadata,
+    now_ms: u64,
+    tolerance_ms: u64,
+) -> Result<(), Error> {
+    let metadata_time = metadata.time_ms;
+
+    // metadata_time - tolerance_ms <= now_ms <= metadata_time + tolerance_ms
+    if now_ms.abs_diff(metadata_time) > tolerance_ms {
```
[the remark this answers is not in the document]
Comment (lklimek) [reply packages/rs-sdk/src/sdk.rs:585]: We use rust 1.80

File: packages/rs-sdk/src/sdk.rs [hunk 9]
```diff
@@ -544,6 +567,105 @@ impl Sdk {
...
+        received_height = received_height,
+        tolerance,
+        "received message with new height"
+    );
+    while let Err(stored_height) = last_seen_height.compare_exchange(
+        expected_height,
+        received_height,
+        Ordering::SeqCst,
+        Ordering::Relaxed,
+    ) {
```
[the remark this answers is not in the document]
Comment (lklimek) [reply packages/rs-sdk/src/sdk.rs:658]: failure ordering doesn't matter in this case, as receiving old data will in pessimistic scenario trigger one loop execution more.

**Review by shumkov — APPROVED** [review]
👍

**Pull request merged**